crates/timers/src/callback.rs

```diff
@@ -1,8 +1,50 @@
 //! Callback-style timer APIs.
 
+use js_sys::{Function, Reflect};
 use wasm_bindgen::prelude::*;
-use wasm_bindgen::JsCast;
-use web_sys::window;
+use wasm_bindgen::{throw_val, JsCast, JsValue};
+use web_sys::{Window, WorkerGlobalScope};
+
+thread_local! {
+    static GLOBAL: WindowOrWorker = WindowOrWorker::new();
+}
+
+enum WindowOrWorker {
+    Window(Window),
+    Worker(WorkerGlobalScope),
+}
+
+impl WindowOrWorker {
+    fn new() -> Self {
+        let global: JsValue = js_sys::global().into();
+
+        if Reflect::has(&global, &String::from("Window").into()).unwrap_throw() {
+            WindowOrWorker::Window(global.into())
+        } else if Reflect::has(&global, &String::from("WorkerGlobalScope").into()).unwrap_throw() {
+            WindowOrWorker::Worker(global.into())
+        } else {
+            throw_val(global)
+        }
+    }
+}
+
+macro_rules! impl_window_or_worker {
+    ($name:ident, ($($par_name:ident: $par_type:ty),*)$(, $return:ty)?) => {
```

Comment:
I think this would be better like this:

```rust
macro_rules! impl_window_or_worker {
    ($(fn $name:ident($($par_name:ident: $par_type:ty),*)$( -> $return:ty)?;)+) => {
        impl WindowOrWorker {
            $(
                fn $name(&self, $($par_name: $par_type),*)$( -> $return)? {
                    match self {
                        Self::Window(window) => window.$name($($par_name),*),
                        Self::Worker(worker) => worker.$name($($par_name),*),
                    }
                }
            )+
        }
    };
}

impl_window_or_worker! {
    fn set_timeout_with_callback_and_timeout_and_arguments_0(handler: &Function, timeout: i32) -> Result<i32, JsValue>;
    fn set_interval_with_callback_and_timeout_and_arguments_0(handler: &Function, timeout: i32) -> Result<i32, JsValue>;
    fn clear_timeout_with_handle(handle: i32);
    fn clear_interval_with_handle(handle: i32);
}
```

Reply:
Sure, I just looked over it and copy pasted.